File: 0094 - Binary Tree Inorder Traversal/Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        self.dfs(root, result)
        return result
        
    def dfs(self, root: TreeNode, result: List[int]) -> None:
        if root:
            if root.left is not None:
                self.dfs(root.left, result)
            result.append(root.val)
            if root.right is not None:
                self.dfs(root.right, result)

File: 0094 - Binary Tree Inorder Traversal/Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        dfs(root, result);
        return result;
    }
    
    void dfs(TreeNode* root, vector<int>& result) {
        if (root != NULL) {
            if (root->left != NULL) {
                dfs(root->left, result);
            }
            result.push_back(root->val);
            if (root->right != NULL) {
                dfs(root->right, result);
            }
        }
    }
};
